File: Programmers/순위검색.py
from bisect import bisect_left
info = ["java backend junior pizza 150", "python frontend senior chicken 210", "python frontend senior chicken 150",
        "cpp backend senior pizza 260", "java backend junior chicken 80", "python backend senior chicken 50"]
query = ["java and backend and junior and pizza 100", "python and frontend and senior and chicken 200",
         "cpp and - and senior and pizza 250", "- and backend and senior and - 150", "- and - and - and chicken 100", "- and - and - and - 150"]

dic = {}
for l in ["cpp", "java", 'python', '-']:
    for g in ['frontend', 'backend', '-']:
        for e in ["junior", "senior", '-']:
            for f in ['pizza', 'chicken', '-']:
                dic[l+g+e+f] = []
for x in range(len(info)):
    lang, group, exp, food, point = list(info[x].split())
    for l in [lang, '-']:
        for g in [group, '-']:
            for e in [exp, '-']:
                for f in [food, '-']:
                    dic[l+g+e+f].append(int(point))
for i in dic:
    if dic[i] != []:
        dic[i].sort()

answer = []
for now in query:
    l, g, e, f = (now.split(' and '))
    f, p = f.split()
    now = dic[l+g+e+f]
    pivot = len(now)
    i = bisect_left(now, int(p))
    answer.append(pivot-i)
print(answer)

# lang = {"java": set(), "cpp": set(), "python": set()}
# exp = {"junior": set(), "senior": set()}
# food = {"pizza": set(), "chicken": set()}
# group = {"backend": set(), "frontend": set()}
# point = []

# for x in range(len(info)):
#     l, g, e, f, p = list(info[x].split())
#     lang[l].add(x)
#     group[g].add(x)
#     exp[e].add(x)
#     food[f].add(x)
#     point.append([int(p), x])
# point.sort()
# Pindex = [i[0] for i in point]
# Iindex = [i[1] for i in point]
# answer = []
# for now in query:
#     l, g, e, f = (now.split(' and '))
#     f, p = f.split()
#     newIndex = bisect_left(Pindex, int(p))
#     pivot = set(Iindex[newIndex:])
#     if l != '-':
#         pivot = pivot & lang[l]
#     if g != '-':
#         pivot = pivot & group[g]
#     if e != '-':
#         pivot = pivot & exp[e]
#     if f != '-':
#         pivot = pivot & food[f]
#     answer.append(len(pivot))
# print(answer)


'''
조건이 간단하다면 브루트포스 한 방법이 더 나은 방법일 수 있다.
'''

File: Programmers/순위검색.js
const cl = console.log;

function solution(info, query) {
  const BS = (array, target) => {
    var left = 0;
    var right = array.length - 1;
    if (target <= array[0]) {
      return 0;
    } else if (target > array[right]) {
      return false;
    } else if (array.length === 0) {
      return false;
    }
    while (left <= right) {
      if (right - left == 1) {
        if (array[left] === target) {
          return left;
        }
        return left + 1;
      }
      var mid = Math.floor((right + left) / 2);
      if (array[mid] >= target) {
        right = mid;
      } else {
        left = mid;
      }
    }
  };

  var answer = [];
  var dic = {};
  const newInfo = Array(info)[0];
  const Q = Array(query)[0];

  for (lang of ["java", "python", "cpp", "-"]) {
    for (group of ["frontend", "backend", "-"]) {
      for (career of ["junior", "senior", "-"]) {
        for (food of ["chicken", "pizza", "-"])
          dic[lang + group + career + food] = [];
      }
    }
  }

  for (item of newInfo) {
    const I = item.split(" ").filter((item) => item !== "and");
    for (a of ["-", I[0]]) {
      for (b of ["-", I[1]]) {
        for (c of ["-", I[2]]) {
          for (d of ["-", I[3]]) {
            dic[a + b + c + d].push(+I[4]);
          }
        }
      }
    }
  }

  for (lang of ["java", "python", "cpp", "-"]) {
    for (group of ["frontend", "backend", "-"]) {
      for (career of ["junior", "senior", "-"]) {
        for (food of ["chicken", "pizza", "-"])
          dic[lang + group + career + food].sort((a, b) => a - b);
      }
    }
  }

  for (item of Q) {
    const I = item.split(" ").filter((item) => item !== "and");
    const node = BS(dic[I[0] + I[1] + I[2] + I[3]], +I[4]);
    cl(node, dic[I[0] + I[1] + I[2] + I[3]].length);
    node !== false
      ? answer.push(dic[I[0] + I[1] + I[2] + I[3]].length - node)
      : answer.push(0);
  }

  return answer;
}

solution(
  [
    "java backend junior pizza 150",
    "python frontend senior chicken 210",
    "python frontend senior chicken 150",
    "cpp backend senior pizza 260",
    "java backend junior chicken 80",
    "python backend senior chicken 50",
  ],
  [
    "java and backend and junior and pizza 100",
    "python and frontend and senior and chicken 200",
    "cpp and - and senior and pizza 250",
    "cpp and frontend and senior and pizza 250",
    "- and backend and senior and - 150",
    "- and - and - and chicken 100",
    "- and - and - and - 150",
  ]
);
